x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.test.ts

```diff
@@ -98,6 +98,89 @@ describe('waterfall_helpers', () => {
         getWaterfallItems(childrenByParentId, entryTransactionItem)
       ).toMatchSnapshot();
     });
+
+    it('should handle cyclic references', () => {
+      const items: IWaterfallItem[] = [
+        {
+          id: 'c',
+          parentId: 'c',
+          serviceName: 'opbeans-java',
+          name: 'SELECT',
+          duration: 210,
+          timestamp: 1536763736371000,
+          offset: 0,
+          skew: 0,
+          docType: 'span',
+          span: {
+            transaction: {
+              id: 'c'
+            }
+          } as Span
+        },
+        {
+          id: 'b',
+          parentId: 'a',
+          serviceName: 'opbeans-java',
+          name: 'GET [0:0:0:0:0:0:0:1]',
+          duration: 4694,
+          timestamp: 1536763736368000,
+          offset: 0,
+          skew: 0,
+          docType: 'span',
+          span: {
+            transaction: {
+              id: 'a'
+            }
+          } as Span
+        },
+        {
+          id: 'b2',
+          parentId: 'a',
+          serviceName: 'opbeans-java',
+          name: 'GET [0:0:0:0:0:0:0:1]',
+          duration: 4694,
+          timestamp: 1536763736367000,
+          offset: 0,
+          skew: 0,
+          docType: 'span',
+          span: {
+            transaction: {
+              id: 'a'
+            }
+          } as Span
+        },
+        {
+          id: 'c',
+          parentId: 'b',
+          serviceName: 'opbeans-java',
+          name: 'APIRestController#productsRemote',
+          duration: 3581,
+          timestamp: 1536763736369000,
+          offset: 0,
+          skew: 0,
+          docType: 'transaction',
+          transaction: {} as Transaction
+        },
+        {
+          id: 'a',
+          serviceName: 'opbeans-java',
+          name: 'APIRestController#products',
+          duration: 9480,
+          timestamp: 1536763736366000,
+          offset: 0,
+          skew: 0,
+          docType: 'transaction',
+          transaction: {} as Transaction
+        }
+      ];
```

Comment:
Would it be possible to trim the `items` down to the essentials, so it's easier to read the test, and grok what it's doing?

This will still create the recursive problem - do you think it's too short?
```js
      const items = [
        { id: 'a', timestamp: 10 } as IWaterfallItem,
        { id: 'a', parentId: 'a', timestamp: 20 } as IWaterfallItem
      ];
```

Reply:
that makes sense